src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -44,16 +87,227 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+}
+void thread_server(void *arg) {
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: In thread_server\n", tid);
+    goto_work(NULL);
+    FARF(LOW, "HVX_TP: %d: thread_server: Exiting with QURT_EOK.\n", tid);
+    qurt_thread_exit(QURT_EOK);
+}
+void initialize_work_queue() {
+    FARF(LOW, "HVX_TP: Master Thread: Initializing work queue\n");
+    lock_work_queue();
+    work_queue.shutdown = false;
+    work_queue.jobs = NULL;
+    FARF(LOW, "HVX_TP: Master Thread: Work queue initialized\n");
+    unlock_work_queue();
+}
+void create_threads(int num_threads) {
+    // Acquire a lock on the work queue.
+    FARF(LOW, "HVX_TP: Master Thread: Creating Threads\n");
+
+    int i;
+    qurt_thread_attr_t thread_attr;
+    for (i = 0; i < num_threads; ++i) {
+        qurt_thread_attr_init(&thread_attr);
+        qurt_thread_attr_set_stack_addr(&thread_attr, stack[i]);
+        qurt_thread_attr_set_stack_size(&thread_attr, 1024);
+        qurt_thread_attr_set_priority(&thread_attr, 100);
+        qurt_thread_create(&threads[i], &thread_attr, thread_server, (void *)i);
+    }
+    FARF(LOW, "HVX_TP: Master Thread: Created threads\n");
+}
+void qurt_thread_pool_init() {
+    FARF(LOW, "HVX_TP: Master Thread: Initializing the thread pool\n");
+    qurt_mutex_init(&(work_queue.work_mutex));
+    qurt_sem_init_val(&wait_for_work, 0);
+
+    initialize_work_queue();
+
+    create_threads(NUM_WORKER_THREADS_TO_CREATE);
+
+    thread_pool_initialized = true;
+    FARF(LOW, "HVX_TP: Master Thread: Thread pool initialized\n");
+}
+
+int halide_do_par_for(void *user_context, halide_task_t f,
+                      int min, int size, uint8_t *closure) {
+    FARF(LOW, "HVX_TP: Master Thread: halide_do_par_for\n");
+
+    // 1. If the thread pool hasn't been initialized, initiliaze it.
+    // This involves.
+    //    a) Creating the threads.
+    //    b) Acquiring a lock on the work queue and clearing the jobs therein
+    //    c) Setting up a semaphore on which worker threads sleep until awoken
+    //       by this thread i.e. the master thread.
+    if (!thread_pool_initialized) {
+        // Initialize the work queue mutex.
+        // Initialize the wait_for_work semaphore.
+        // lock work queue
+        //    wq.shutdown = false;
+        //    wq.jobs = NULL;
+        // unlcok work queue
+        // create NUM_WORKER_THREADS_TO_CREATE number of threads.
+        // thread_pool_initialized = true;
+        qurt_thread_pool_init();
+    }
+
+    // 2. Lock the work queue again.
+    lock_work_queue();
+
+    // 3. Put work in the global work queue.
+
+    FARF(LOW, "HVX_TP: Master Thread: Putting job in the work queue at  min = %d, size = %d, user_context = 0x%x, f= 0x%x, \n",
+             min, size, user_context, f);
+
+    work job;
+    job.f = f;
+    job.user_context = user_context;
+    job.next = min;          // Start at this index.
+    job.max  = min + size;   // Keep going until one less than this index.
```

Comment:
The value currently named 'max' seems more like what the STL would call 'end' (one past the last index).